### PR TITLE
gnome-backgrounds: stable image paths

### PR DESCRIPTION
**Pull request by cwyc**
###### Motivation for this change
The background database located in $out/share/gnome-background-properties/gnome-backgrounds.xml points to image paths in the /nix/store.
When a background is selected in gnome-control-center, its path at that moment is stored in a dconf setting.
It isn't updated when this package's hash changes, so once the old path is garbage-collected, gnome can no longer find the file.
So, we need to make it point to the stable directory /run/current-system/sw/share/backgrounds/gnome.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/desktops/gnome/core/gnome-backgrounds/default.nix [hunk 1]
```diff
@@ -13,6 +13,9 @@ stdenv.mkDerivation rec {
     updateScript = gnome.updateScript { packageName = "gnome-backgrounds"; attrPath = "gnome.gnome-backgrounds"; };
   };
 
+  # Makes the database point to stable paths in /run/current-system/sw that don't decay whenever this package's hash changes.
+  patches = [ ./stable-dir.patch ];
```
Comment (jtojnar) [inline pkgs/desktops/gnome/core/gnome-backgrounds/default.nix:17]:
```suggestion
  patches = [
    # Makes the database point to stable paths in /run/current-system/sw that don't decay whenever this package's hash changes.
    ./stable-dir.patch
  ];
```

**Comment by jtojnar** [comment]
We also need to make sure that gnome-backgrounds is installed to `environment.systemPackages` (probably should add a comment that the settings depend on it being installed globally next to the line in the NixOS module) and that the directory containing backgrounds is listed in `environment.pathsToLink` (in accordance with https://github.com/NixOS/nixpkgs/issues/47173).

**Comment by cwyc** [comment]
How about this?
```nix
    # Makes the database point to stable paths in /run/current-system/sw/share, which don't decay whenever this package's hash changes.
    # This assumes a nixos + gnome system, where this package is installed globally, and /share outputs are linked.
```
Also, would a option to enable/disable this patch be necessary? It's only relevant for nixos/gnome systems, but I don't imagine it matters for any other platform.

**Comment by jtojnar** [comment]
Yeah, I do not think we need to worry about anything but NixOS here. But I think it still makes sense to include the comment in the module so I do not accidentally remove it in the future. And the explicit `pathsToLink` for the same reason.

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 134509` at 7dc8bda7 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>gnome.gnome-backgrounds</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/license-missing.md">license-missing</a>

Package is missing a license.

Near pkgs/desktops/gnome/core/gnome-backgrounds/default.nix:25:3:
```
   |
25 |   meta = with lib; {
   |   ^
```
</li>
  </ul>
</details>

**Review by amaxine — COMMENTED** [review]
Maybe a silly question, but can we simply set `datadir` to that path as part of the options, instead of patching?

**Comment by cwyc** [comment]
@maxeaubrey in [backgrounds/build.meson](https://gitlab.gnome.org/GNOME/gnome-backgrounds/-/blob/master/backgrounds/meson.build#L29) it installs the files in backgrounddir and metadatadir, which are derived from datadir, so datadir still has to point to $out.

**Pull request merged**